### PR TITLE
Coordinates

### PR DESCRIPTION
**Pull request by Jammy2211**
A Coordinates object is a structured list of (y,x) values, e.g:

[[(1.0, 1.0)], [(2.0, 2.0), (3.0, 3.0)]]

The structure represents that coordinates come from a shared origin. For example, in autolens, (1.0, 1.0) may correspond to the centre of a mass profile of galaxy g0, and (2.0, 2.0), (3.0, 3.0) the centres of the mass profiles of galaxy g1.

This distinguishes them from a Grid object, which cannot be 'grouped' in this way. Furthermore, a Grid object is defined to originate from a uniform grid of (y,x) coordinates, albeit the values may no longer be uniform due to lensing.

We want all _from_grid methods to accept both a Grid object and Coordinates object, so a use can, for example, determine the deflectioon angles at all coordinates on a uniform grid or at an irregular set of locations. Thus, both Grid and Coordinate objets store values as a 2D NumPy array of shape [total_coordinates, 2] which is passed to these functions. The grid_like_to_numpy decrotator handles the conversions to / from these different data structures when performing a calculations.

I have now also made a *Values* class, which structures things the same as Coordinates but for floats, e.g.:

[[1.0], [2.0, 3.0]]

This means from_grid methods that return scalars can retain the same structure as the input Coordinates.

### CODE REVIEW

File: autoarray/structures/arrays.py [hunk 1]
```diff
@@ -478,3 +480,194 @@ def zeros(cls, mask, store_in_1d=True):
...
+        calculations.
+
+        The values input to this function can have any of the following forms:
+
+        [[x0, x1], [x0]]
+        [[[x0, x1]], [x0]]
+
+        In all cases, they will be converted to a list of list of floats followed by a 1D NumPy array.
+
+        Print methods are overridden so a user always "sees" the values as the list structure.
```
Comment (rhayes777) [inline autoarray/structures/arrays.py:508]: Do you mean __str__ and __repr__?

File: autoarray/structures/arrays.py [hunk 2]
```diff
@@ -478,3 +480,194 @@ def zeros(cls, mask, store_in_1d=True):
...
+
+        Parameters
+        ----------
+        values : [[float]] or equivalent
+            A collection of values that are grouped if they correpsond to a shared origin.
+        mask : aa.Mask
+            The mask whose attributes are used to perform coordinate conversions.
+        """
+
+        if values == []:
```
Comment (rhayes777) [inline autoarray/structures/arrays.py:521]: if len(values) == 0

File: autoarray/structures/arrays.py [hunk 3]
```diff
@@ -478,3 +480,194 @@ def zeros(cls, mask, store_in_1d=True):
...
+
+class Values(np.ndarray):
+    def __new__(cls, values, mask=None):
+        """ A collection of values structured in a way defining groups of values which share a common origin (for
+        example values may be grouped if they are from a specific region of a dataset).
+
+        Grouping is structured as follows:
+
+        [[(y0,x0), (y1,x1)], [(y0,x0), (y1,x1), (y2,x2)]]
+
```
Comment (rhayes777) [inline autoarray/structures/arrays.py:493]: Might be worth explaining the use of this class

File: autoarray/structures/arrays.py [hunk 4]
```diff
@@ -478,3 +480,194 @@ def zeros(cls, mask, store_in_1d=True):
...
+            return []
+
+        if isinstance(values[0], float):
+            values = [values]
+
+        upper_indexes = []
+
+        a = 0
+
+        for coords in values:
```
Comment (rhayes777) [inline autoarray/structures/arrays.py:531]: a = sum(map(len, values))

File: autoarray/structures/arrays.py [hunk 5]
```diff
@@ -478,3 +480,194 @@ def zeros(cls, mask, store_in_1d=True):
...
+            self.lower_indexes = obj.lower_indexes
+
+        if hasattr(obj, "upper_indexes"):
+            self.upper_indexes = obj.upper_indexes
+
+        if hasattr(obj, "mask"):
+            self.mask = obj.mask
+
+    @classmethod
+    def from_pixels_and_mask(cls, pixels, mask):
```
Comment (rhayes777) [inline autoarray/structures/arrays.py:556]: Try adding type annotations

File: autoarray/structures/arrays.py [hunk 6]
```diff
@@ -478,3 +480,194 @@ def zeros(cls, mask, store_in_1d=True):
...
+        return self
+
+    @property
+    def in_list(self):
+        """Return the values on a structured list which groups values with a common origin."""
+        return [list(self[i:j]) for i, j in zip(self.lower_indexes, self.upper_indexes)]
+
+    @property
+    def in_list_1d(self):
+        """Return the values in a list without structured grouping."""
```
Comment (rhayes777) [inline autoarray/structures/arrays.py:580]: I'd be inclined to let the user cast to a list. This function doesn't do much

File: autoarray/structures/arrays.py [hunk 7]
```diff
@@ -478,3 +480,194 @@ def zeros(cls, mask, store_in_1d=True):
...
+        return list(self)
+
+    def values_from_arr_1d(self, arr_1d):
+        """Create a *Values* object from a 1D NumPy array of values of shape [total_values]. The
+        *Values* are structured and grouped following this *Coordinate* instance."""
+        values_1d = [
+            list(arr_1d[i:j]) for i, j in zip(self.lower_indexes, self.upper_indexes)
```
Comment (rhayes777) [inline autoarray/structures/arrays.py:587]: in_list should call this function to avoid repetition

File: autoarray/structures/arrays.py [hunk 8]
```diff
@@ -478,3 +480,194 @@ def zeros(cls, mask, store_in_1d=True):
...
+    def from_file(cls, file_path):
+        """Create a *Coordinates* object from a file which stores the values as a list of list of tuples.
+
+        Parameters
+        ----------
+        file_path : str
+            The path to the values .dat file containing the values (e.g. '/path/to/values.dat')
+        """
+        with open(file_path) as f:
+            values_string = f.readlines()
```
Comment (rhayes777) [inline autoarray/structures/arrays.py:611]: This "string" is not a string. Needs to be renamed

File: autoarray/structures/arrays.py [hunk 9]
```diff
@@ -478,3 +480,194 @@ def zeros(cls, mask, store_in_1d=True):
...
+
+        Parameters
+        ----------
+        file_path : str
+            The path to the values .dat file containing the values (e.g. '/path/to/values.dat')
+        overwrite : bool
+            If there is as exsiting file it will be overwritten if this is *True*.
+        """
+
+        if overwrite and os.path.exists(file_path):
```
Comment (rhayes777) [inline autoarray/structures/arrays.py:632]: You should call os.path.exists then do an if else inside to simplify the logic

File: autoarray/structures/arrays.py [hunk 10]
```diff
@@ -478,3 +480,194 @@ def zeros(cls, mask, store_in_1d=True):
...
+            os.remove(file_path)
+        elif not overwrite and os.path.exists(file_path):
+            raise FileExistsError(
+                "The file ",
```
Comment (rhayes777) [inline autoarray/structures/arrays.py:636]: Use an f string

File: autoarray/structures/arrays.py [hunk 11]
```diff
@@ -478,3 +480,194 @@ def zeros(cls, mask, store_in_1d=True):
...
+                file_path,
+                " already exists. Set overwrite=True to overwrite this" "file",
+            )
+
+        with open(file_path, "w") as f:
```
Comment (rhayes777) [inline autoarray/structures/arrays.py:641]: So this actually won't work unless the file exists. "w+" is required to create the file. That seems to contradict your overwrite behaviour?

File: autoarray/structures/arrays.py [hunk 12]
```diff
@@ -478,3 +480,194 @@ def zeros(cls, mask, store_in_1d=True):
...
+            for value in self.in_list:
+                f.write("%s\n" % value)
+
+    @property
+    def mapping(self):
+        return self.mask.mapping
```
Comment (rhayes777) [inline autoarray/structures/arrays.py:647]: Is it so hard to call .mask.mapping? I'd only bother with this method if you have to conform to the same interface

File: autoarray/structures/arrays.py [hunk 13]
```diff
@@ -478,3 +480,194 @@ def zeros(cls, mask, store_in_1d=True):
...
+
+    @property
+    def shape_2d_scaled(self):
+        return (
+            np.amax(self.in_1d[:, 0]) - np.amin(self.in_1d[:, 0]),
```
Comment (rhayes777) [inline autoarray/structures/arrays.py:652]: You seem to have forgotten about docs down here

File: autoarray/structures/arrays.py [hunk 14]
```diff
@@ -478,3 +480,194 @@ def zeros(cls, mask, store_in_1d=True):
...
+            np.amax(self.in_1d[:, 1]) - np.amin(self.in_1d[:, 1]),
+        )
+
+    @property
+    def scaled_maxima(self):
+        return (np.amax(self.in_1d[:, 0]), np.amax(self.in_1d[:, 1]))
```
Comment (rhayes777) [inline autoarray/structures/arrays.py:658]: These brackets are unnecessary. I'm sure if you go look at pycharm the linter will tell you that

File: autoarray/structures/grids.py [hunk 15]
```diff
@@ -751,7 +755,135 @@ def blurring_grid_from_mask_and_kernel_shape(
             )
 
 
-class GridIrregular(np.ndarray):
+class GridRectangular(Grid):
+    def __new__(cls, grid, shape_2d, pixel_scales, origin=(0.0, 0.0), *args, **kwargs):
```
Comment (rhayes777) [inline autoarray/structures/grids.py:759]: I'd probably new-line these arguments as it's getting quite long

File: autoarray/structures/grids.py [hunk 16]
```diff
@@ -751,7 +755,135 @@ def blurring_grid_from_mask_and_kernel_shape(
...
+        obj = super().__new__(cls=cls, grid=grid, mask=mask)
+
+        pixel_neighbors, pixel_neighbors_size = pixelization_util.rectangular_neighbors_from_shape(
+            shape=shape_2d
+        )
+        obj.pixel_neighbors = pixel_neighbors.astype("int")
+        obj.pixel_neighbors_size = pixel_neighbors_size.astype("int")
+        return obj
+
+    def __array_finalize__(self, obj):
```
Comment (rhayes777) [inline autoarray/structures/grids.py:792]: Does this need to be here?

File: autoarray/structures/grids.py [hunk 17]
```diff
@@ -751,7 +755,135 @@ def blurring_grid_from_mask_and_kernel_shape(
...
+        pass
+
+    def __reduce__(self):
+        # Get the parent's __reduce__ tuple
+        pickled_state = super(GridRectangular, self).__reduce__()
```
Comment (rhayes777) [inline autoarray/structures/grids.py:797]: No need for arguments in super call

File: autoarray/structures/grids.py [hunk 18]
```diff
@@ -751,7 +755,135 @@ def blurring_grid_from_mask_and_kernel_shape(
...
+        new_state = pickled_state[2] + (class_dict,)
+        # Return a tuple that replaces the parent's __setstate__ tuple with our own
+        return pickled_state[0], pickled_state[1], new_state
+
+    # noinspection PyMethodOverriding
+    def __setstate__(self, state):
+
+        for key, value in state[-1].items():
+            setattr(self, key, value)
+        super(GridRectangular, self).__setstate__(state[0:-1])
```
Comment (rhayes777) [inline autoarray/structures/grids.py:811]: Same here

File: autoarray/structures/grids.py [hunk 19]
```diff
@@ -773,26 +905,53 @@ def __new__(
         nearest_pixelization_1d_index_for_mask_1d_index : ndarray
             A 1D array that maps every grid pixel to its nearest pixelization-grid pixel.
         """
+
         if type(grid) is list:
```
Comment (rhayes777) [inline autoarray/structures/grids.py:909]: if isinstance(grid, list)

File: autoarray/structures/grids.py [hunk 20]
```diff
@@ -806,15 +965,11 @@ def __setstate__(self, state):
 
         for key, value in state[-1].items():
             setattr(self, key, value)
-        super(GridIrregular, self).__setstate__(state[0:-1])
+        super(GridVoronoi, self).__setstate__(state[0:-1])
```
Comment (rhayes777) [inline autoarray/structures/grids.py:968]: No need to pass arguments to super call

File: autoarray/structures/grids.py [hunk 21]
```diff
@@ -823,53 +978,11 @@ def from_grid_and_unmasked_2d_grid_shape(cls, unmasked_sparse_shape, grid):
             unmasked_sparse_shape=unmasked_sparse_shape, grid=grid
         )
 
-        return GridIrregular(
+        return GridVoronoi(
             grid=sparse_grid.sparse,
             nearest_pixelization_1d_index_for_mask_1d_index=sparse_grid.sparse_1d_index_for_mask_1d_index,
```
Comment (rhayes777) [inline autoarray/structures/grids.py:983]: quite a long argument name

File: autoarray/structures/grids.py [hunk 22]
```diff
@@ -1317,93 +1220,142 @@ def interpolated_values_from_values(self, values):
...
+            The mask whose attributes are used to perform coordinate conversions.
+        """
 
-        coordinates_list = []
+        if len(coordinates) == 0:
+            return []
 
-        for coordinate_set in self:
-            coordinates_list.append(GridIrregular(grid=coordinate_set))
+        if isinstance(coordinates[0], tuple) or isinstance(coordinates[0], np.ndarray):
```
Comment (rhayes777) [inline autoarray/structures/grids.py:1264]: if isinstance(coordinates[0], (tuple, np.ndarray))

File: autoarray/structures/grids.py [hunk 23]
```diff
@@ -1317,93 +1220,142 @@ def interpolated_values_from_values(self, values):
...
+    def in_1d(self):
         return self
 
     @property
-    def pixels(self):
+    def in_list(self):
+        """Return the coordinates on a structured list which groups coordinates with a common origin."""
+        return [
+            list(map(tuple, self[i:j, :]))
+            for i, j in zip(self.lower_indexes, self.upper_indexes)
```
Comment (rhayes777) [inline autoarray/structures/grids.py:1329]: There's a lot of repetition between these methods and those in values. Might make sense to have a common mixin to implement these methods?

File: autoarray/structures/grids.py [hunk 24]
```diff
@@ -1471,11 +1407,68 @@ def output_to_file(self, file_path, overwrite=False):
             )
 
         with open(file_path, "w") as f:
-            for coordinate in self:
+            for coordinate in self.in_list:
                 f.write("%s\n" % coordinate)
```
Comment (rhayes777) [inline autoarray/structures/grids.py:1411]: Oh my god that's some ancient formatting syntax right there.

use fstrings

File: autoarray/structures/grids.py [hunk 25]
```diff
@@ -1471,11 +1407,68 @@ def output_to_file(self, file_path, overwrite=False):
...
+            [
+                self.scaled_minima[1],
+                self.scaled_maxima[1],
+                self.scaled_minima[0],
+                self.scaled_maxima[0],
+            ]
+        )
+
+
+def grid_like_to_numpy(func):
```
Comment (rhayes777) [inline autoarray/structures/grids.py:1471]: Kind of concerned we're going to see a lot of this decorator.

File: autoarray/structures/grids.py (hunk 23 above)
Comment (Jammy2211) [reply autoarray/structures/grids.py:1329]: Agreed, I am expecting this to all end up looking like that.

**Pull request merged**